### PR TITLE
wxGUI/datacatalog: remove unnecessary check and warning for dragging mapsets/locations

### LINKED ISSUE #1299: [Bug] GUI: Move or copy location/mapset isn't allowed
**Describe the bug**
Whenever I click a location or mapset (even currently selected or the current) from the Data catalog, I'm getting a warning: "Move or copy location isn't allowed" or "Move or copy mapset isn't allowed".

**To Reproduce**

1. Go to 'Data'
2. Click on any location or mapset
3. See error

**Expected behavior**
No warnings or errors when simply clicking a location or mapset. I'm not trying to "move or copy" anything.

**Screenshots**

image

**System description (please complete the following information):**

- Operating System: Linux
- GRASS GIS version: trunk

**Additional context**
Add any other context about the problem here. N/A

### PR DESCRIPTION
**Pull request by petrasovaa**
Should fix #1299

### CODE REVIEW

**Review by nilason — APPROVED** [review]
Great! This fixes the problems for me.
Thanks!!

**Comment by HuidaeCho** [comment]
The reported issue is fixed. Clicking on a location/mapset and moving the cursor doesn't show any warning now. Does it mean we lost dragging to move or copy mapsets?

Also, I found another related issue. I think it was already there. Clicking on a layer and clicking on a non-current mapset triggers a dragging event.
1. Click on a layer (button released)
2. I start seeing a rectangle around mapsets and locations (I'm not dragging it)
3. Click on another mapset or location
4. Invalid warning

Click a location
image

Click a mapset
image

**Comment by HuidaeCho** [comment]
BTW, how can I activate Delete/Rename mapset? They are always deactivated.

**Comment by HuidaeCho** [comment]
I just wonder why mouse move (no buttons pushed and move) is treated as drag (push/hold the left button and move). Even the function name is OnBeginDrag.

**Comment by HuidaeCho** [comment]
Hmm, interesting... I tested it *without* this PR on my laptop and desktop. I had this issue only on my laptop. I ssh'ed into the desktop and experienced the same issue from the laptop remotely, so it doesn't seem to be our code. Hooking up a USB mouse to the laptop didn't help. I'm using the same gpm settings `gpm -m /dev/mouse -t imps1` on both machines.

* Laptop: ThinkPad Yoga 370 with a touch screen and a stylus pen, Slackware 14.2 with the latest slack packages
* Desktop: ThinkStation S20, Slackware 14.2 with the latest slack packages
* wxPython 4.0.7

I keep my touch pad disabled and use the track point.
1. Enabled the touch pad, the issue is still there using the track point/mouse/stylus.
2. Used the touch pad once, the issue is gone for track point/mouse.
3. Using the stylus immediately caused this issue again.
4. Touched the screen, using the track point or mouse caused it.
5. Used the touch pad again, the issue is gone.
6. Just touching the screen was OK.

So.. my conclusion is
* This is not a GRASS issue. Please do NOT merge this PR.
* Touching and releasing the screen triggers a dragging event *only when* the track point or mouse is used next.
* The track point and mouse honor the previous touch and start dragging, but the touch pad resets it.
* Touching the screen with the stylus and hovering it triggers a dragging event.
* It can be an issue with the [Synaptics TouchPad driver](http://www.thinkwiki.org/wiki/Synaptics_TouchPad_driver_for_X) or [xinput](http://www.thinkwiki.org/wiki/How_to_configure_the_TrackPoint).

Dragging on ThinkPad
* touch pad + left button + move
* touch the screen + track point/mouse with no button pushed + move
* touch the stylus + hover it

Cancelling a drag event triggered by touching the screen
* touch pad

Basically, touching the screen is push and hold the left button, any *continuous* move afterward is treated as dragging, and touching the touch pad releases it. Finger touching doesn't generate continuous move events because there is no tracing. Swiping the touch pad without a button pressed is not dragging.

@nilason Are you using a laptop with a touch screen?

**Comment by wenzeslaus** [comment]
> This is not a GRASS issue. Please do NOT merge this PR.

I have the same issue too and I want to see this fixed. It seems more like a mouse sensitivity issue for me as I get it sometimes during clicking and more likely during a double click, so I think I move my mouse slightly in the process. Driver or sensitivity, we should not punish the user for misclicking which is ultimately harmless.

Perhaps the case which needs to be ignored is only the one when moving by dragging is done to the same location (BTW, do I mean a place in file structure or a GRASS location? You need to guess!).

In many cases with drag and drop, I think the result should be simply no-op when the action cannot be performed.

**Comment by HuidaeCho** [comment]
> > This is not a GRASS issue. Please do NOT merge this PR.
> 
> I have the same issue too and I want to see this fixed. It seems more like a mouse sensitivity issue for me as I get it sometimes during clicking and more likely during a double click, so I think I move my mouse slightly in the process. Driver or sensitivity, we should not punish the user for misclicking which is ultimately harmless.

Do you have this issue on Windows or Linux? If it's on Windows... well.. we need a fix. Maybe, just no dragging feature at all? I don't know...

> 
> Perhaps the case which needs to be ignored is only the one when moving by dragging is done to the same location (BTW, do I mean a place in file structure or a GRASS location? You need to guess!).
> 
> In many cases with drag and drop, I think the result should be simply no-op when the action cannot be performed.

**Comment by wenzeslaus** [comment]
> Do you have this issue on Windows or Linux? If it's on Windows... well.. we need a fix.

It's Linux (Pop!_OS), System76 laptop.

> Maybe, just no dragging feature at all? I don't know...

That would be throwing the baby out with the bathwater. Intentional drag and drop works fine.

> Also, it sounds like your issue is different from mine. My issue is very predictable and easy to reproduce. Is yours random or does it have a predictable pattern?

I can reproduce it in just few seconds or clicks, but I didn't do all any of the tests you did because it always seemed related to making a small move with the mouse (which has no effect when clicking on buttons or for other things).

**Comment by wenzeslaus** [comment]
> Perhaps the case which needs to be ignored is only the one when moving by dragging is done to the same location...In many cases with drag and drop, I think the result should be simply no-op when the action cannot be performed.

Actually, the error appears on "begin drag" not when finished which is what the function in question is handling, so there is no way to tell what is user's intention. However, when drag and drop is not possible at all, standard behavior simply is that no drag is even started, so this goes back to the code in this PR: When it is not possible to do anything, simply deny/veto the event, i.e., move on.

**Comment by petrasovaa** [comment]
I would merge this fix, all it's doing is not showing a message dialog. So the only downside is that users who actually want to drag a mapset won't be explicitely informed that they are not supposed to do it, but the result is the same. It happened to me couple times that I unintentionally started to drag mapset and this dialog was annoying.

**Comment by HuidaeCho** [comment]
OK, I have never used this dragging feature, so I have no objection to merging it if it works for you.

* Before: warning at the beginning of dragging
* This PR: no warning at all even at the end of dragging? Just silent failure? Is it possible to display this warning when releasing the button?

**Comment by HuidaeCho** [comment]
> I would merge this fix, all it's doing is not showing a message dialog. So the only downside is that users who actually want to drag a mapset won't be explicitely informed that they are not supposed to do it, but the result is the same. It happened to me couple times that I unintentionally started to drag mapset and this dialog was annoying.

Meaning dragging a mapset is not supported even within the same location? Then what is this dragging function for?

**Comment by nilason** [comment]
Dragging works as expected, but no annoying warnings, after this PR.
I see **absolutely** no reason not to merge.
I guess no-one is happier with a message like: "*Move or copy mapset isn't allowed. Get yourself new hardware!*".
Maybe this isn't problem with the code in GRASS, but this wouldn't be the first nor last time we have to address/avoid underlying problems.

**Comment by petrasovaa** [comment]
> > I would merge this fix, all it's doing is not showing a message dialog. So the only downside is that users who actually want to drag a mapset won't be explicitely informed that they are not supposed to do it, but the result is the same. It happened to me couple times that I unintentionally started to drag mapset and this dialog was annoying.
> 
> Meaning dragging a mapset is not supported even within the same location? Then what is this dragging function for?

Dragging is just not implemented for mapsets and locations because the implementation would be more difficult (reprojecting entire mapset etc). It does work for layers.

**Comment by petrasovaa** [comment]
> OK, I have never used this dragging feature, so I have no objection to merging it if it works for you.
> 
> * Before: warning at the beginning of dragging
> * This PR: no warning at all even at the end of dragging? Just silent failure? Is it possible to display this warning when releasing the button?

Right now the dragging event is cancelled right away, so there is no 'end drag' event.

**Comment by HuidaeCho** [comment]
> The reported issue is fixed. Clicking on a location/mapset and moving the cursor doesn't show any warning now. Does it mean we lost dragging to move or copy mapsets?
> 
> Also, I found another related issue. I think it was already there. Clicking on a layer triggers a dragging event.
> 
>     1. Click on a layer (button released)
> 
>     2. I start seeing a rectangle around mapsets and locations (I'm not dragging the layer)
> 
>     3. Click on another mapset or location
> 
>     4. Invalid warning
> 
> 
> Click a location
> image
> 
> Click a mapset
> image

@petrasovaa Now, the original fix is merged. Could you please take a look at this similar issue? Is it fixable in the same way?

**Comment by nilason** [comment]
@HuidaeCho Did you activate the "Allow edits outside of the current mapset" button? (Unfortunately also with a pencil icon.)

Screenshot 2021-02-05 at 17 54 04

**Comment by petrasovaa** [comment]
> > Click a location
> > image
> > Click a mapset
> > image
> 
> @petrasovaa Now, the original fix is merged. Could you please take a look at this similar issue? Is it fixable in the same way?

These happen when there is end drag event and these should stay there in my opinion. It sounds like you are getting end drag event when you are not supposed to, but that may be the hardware issue you are mentioning.

**Comment by HuidaeCho** [comment]
> > > Click a location
> > > image
> > > Click a mapset
> > > image
> > 
> > 
> > @petrasovaa Now, the original fix is merged. Could you please take a look at this similar issue? Is it fixable in the same way?
> 
> These happen when there is end drag event and these should stay there in my opinion. It sounds like you are getting end drag event when you are not supposed to, but that may be the hardware issue you are mentioning.

You're right. It's the hardware driver issue indeed. No fix is needed from GRASS.

**Comment by HuidaeCho** [comment]
> @HuidaeCho Did you activate the "Allow edits outside of the c.urrent mapset" button? (Unfortunately also with a pencil icon.)
> 
> Screenshot 2021-02-05 at 17 54 04

Thanks @nilason! That was it.

**Comment by nilason** [comment]
> Thanks @nilason! That was it.

Splendid!
